builder/builder.go

```diff
@@ -0,0 +1,550 @@
+		return b
+	}
+
+	b.variants = append(b.variants, variant{keys: append([]any{key}, keys...)})
+
+	return b
+}
+
+type variant struct {
+	keys    []any
+	pattern []any
+}
+
+func (v *variant) build(spacing string) string {
+	var s string
+
+	for i, k := range v.keys {
+		if i > 0 {
+			s += coalesce(spacing, defaultSpacing)
+		}
+
+		if k == "*" {
+			s += "*"
+		} else {
+			s += printLiteral(k)
+		}
+	}
+
+	s += spacing + "{{"
+
+	for i := range v.pattern {
+		if i > 0 {
+			s += " "
+		}
+
+		switch p := v.pattern[i].(type) {
+		case string:
+			s += textEscape(p)
+		case *Expression:
+			s += p.build(spacing)
+		default:
+			panic(fmt.Sprintf("unsupported pattern type: %T", p))
+		}
+	}
+
+	return s + "}}"
+}
+
+type local struct {
+	expr     *Expression
+	variable variable
+}
+
+type literal any
+
+type variable string
+
+type function struct {
+	name    string
+	options []FuncOption
+}
+
+type Expression struct {
+	operand  any // literal or variable
+	function function
+}
+
+func (e *Expression) build(spacing string) string {
+	s := "{" + spacing
+
+	switch v := e.operand.(type) {
+	case variable:
+		s += string(v)
+	case nil:
+		// noop
+	case literal:
+		s += printLiteral(v)
+	default:
+		panic(fmt.Sprintf("unsupported operand type: %T", v))
+	}
+
+	if e.function.name != "" {
+		if e.operand != nil { // literal or variable
+			s += " "
+		}
+
+		s += e.function.name
+
+		for _, o := range e.function.options {
+			s += " " + o.key + spacing + "=" + spacing
+
+			if v, ok := o.operand.(string); ok && len(v) > 0 && v[0] == '$' { // recognized as variable if string starts with $
+				s += v
+				continue
+			}
+
+			s += printLiteral(o.operand)
+		}
+	}
+
+	return s + spacing + "}"
+}
+
+func Literal(v any) *Expression {
+	return Expr().Literal(v)
+}
+
+func (e *Expression) Literal(v any) *Expression {
+	e.operand = v
+	return e
+}
+
+func Var(s string) *Expression {
+	return Expr().Var(s)
+}
+
+func (e *Expression) Var(v string) *Expression {
+	if len(v) == 0 {
+		panic("variable name cannot be empty")
+	}
+
+	if v[0] != '$' {
+		panic(fmt.Sprintf("variable must start with $: %s", v))
+	}
+
+	e.operand = variable(v)
+
+	return e
+}
+
+type FuncOption struct {
+	operand any // literal or variable
+	key     string
+}
+
+func Option(key string, operand any) FuncOption { return FuncOption{key: key, operand: operand} }
+
+func Func(name string, option ...FuncOption) *Expression {
+	return Expr().Func(name, option...)
+}
+
+func (e *Expression) Func(name string, option ...FuncOption) *Expression {
+	if len(name) == 0 {
+		panic("function name cannot be empty")
+	}
+
+	if e.function.name != "" {
+		panic("expression already has a function")
+	}
+
+	switch name[0] {
+	default:
+		panic(fmt.Sprintf("function MUST start with :, + or -: %s", name))
+	case ':', '+', '-':
+		e.function = function{name: name, options: option}
+		return e
+	}
+}
+
+func printLiteral(l any) string {
+	switch v := l.(type) { // TODO: more liberal
+	case string:
```

Comment:
so 2/3 literal cases are covered
this can print number, and quoted literals, what about unquoted ? it gets ignored ?

Reply:
prints unqouted if `name = name-start *name-char` holds true

Reply:
but i see here only two returns ?
when 3 expected

Reply:
string -> unqouted/quoted
number -> unqouted
unknown -> panic

Reply:
i see now